### PR TITLE
Adds a rule to slug

### PR DESCRIPTION
**Pull request by jhuet**
### What?

Adds a rule to slug a field with its own value or from another field's.
### Checklist
- [x] Added unit test for added/fixed code
- [x] Updated the documentation
- [ ] Scrutinizer code coverage is 100%
- [x] Scrutinizer code quality is as high as possible
### Linked issue

None.

### CODE REVIEW

**Comment by rick-nu** [comment]
Hey! Thank you very much for your contribution! I really like this addition, I will give it a closer look soon, but from what I've seen now it's looking really good. :+1:

**Comment by rick-nu** [comment]
And I will also look into why Scrutinizer isn't building pull requests anymore :\

**Comment by jhuet** [comment]
Hi @RickvdStaaij, i actually have a little problem with its usage right now :)

I think most people will slug a field based on another one, say a `title` for example. Now, if for some reason, i don't provide `title` to the filter, i end up with an empty string `slug` because i used `defaults('')` on `slug` so it would still be filtered if not provided.

To avoid using `defaults('')` i found out about `$allowNotSet = true` property which is a better approach but it will always set `slug` to whatever value i return, even `null`.

What i think should be the desired result would be to actually have no `slug` key in the filtered result at all if its field _provider_ is not being filtered either.

Here's an example for comprehension sake :)

Actual result with `$allowNotSet = true`:

``` php
$filter->value('slug')->slug('name');
$result = $filter->filter([
    'foo' => 'bar',
]);
// $result = array(2) { ["foo"]=> string(3) "bar" ["slug"]=> NULL } 
```

Now the expected result should be :

``` php
$filter->value('slug')->slug('name');
$result = $filter->filter([
    'foo' => 'bar',
]);
// $result = array(2) { ["foo"]=> string(3) "bar" } <- no slug because, no slug provider "name"
```

Any idea on how to acheive this ? Thanks !

**Comment by jhuet** [comment]
Hi @RickvdStaaij, any thought on the question in my last comment ? Thanks !

File: src/FilterResource.php [hunk 1]
```diff
@@ -230,6 +230,16 @@ public function replace($search, $replace)
     }
 
     /**
+     * Results that returns a value slugged
+     *
+     * @param type $fieldToSlugFrom
```
Comment (localheinz) [inline src/FilterResource.php]: @jhuet 

Should this be `string` instead of `type`?

File: src/FilterRule/Slug.php [hunk 2]
```diff
@@ -0,0 +1,67 @@
+<?php
+/**
+ * Particle.
+ *
+ * @link      http://github.com/particle-php for the canonical source repository
+ * @copyright Copyright (c) 2005-2015 Particle (http://particle-php.com)
```
Comment (localheinz) [inline src/FilterRule/Slug.php]: @jhuet @RickvdStaaij 

Should this be `2016` instead of `2015`?

File: tests/FilterRule/SlugTest.php [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    public function testSlugFilterRule($value, $filteredValue, $field, $fieldValue)
+    {
+        $this->filter->value('test')->slug($field);
+
+        $result = $this->filter->filter([
+            'test' => $value,
+            $field => $fieldValue,
+        ]);
+
+        $this->assertEquals($filteredValue, $result['test']);
```
Comment (localheinz) [inline tests/FilterRule/SlugTest.php]: @jhuet 

Can we assert identity here?

**Comment by rick-nu** [comment]
@jhuet Sorry, I haven't had the time yet to check this out, but it's on the top of my todo list now, so I hope to get back to you soon. One thing that comes to mind, have you checkend https://github.com/particle-php/Filter/blob/master/src/Chain.php#L42? If you state that a rule is not empty, it shouldn't be ignored and turn up with the given value.

**Comment by rick-nu** [comment]
Sorry it took me so long @jhuet, but I have (finally) looked into it in https://github.com/jhuet/Filter/pull/1. I introduced an `isEmpty` property on the `Rule`, which can be set to true. This way the value will be ignored when put out of the filter.

I hope this is what you were looking for. If not let me know :)

**Comment by jhuet** [comment]
No problem @RickvdStaaij, i'm glad to add new functionnalities to this library and see you're still around to help if needed :) Thanks !

File: tests/FilterRule/SlugTest.php (hunk 3 above)
Comment (jhuet) [inline tests/FilterRule/SlugTest.php]: Thanks for you review @localheinz, i'm not sure what you mean here though.

Comment (localheinz) [inline tests/FilterRule/SlugTest.php]: @jhuet 

How about this?

``` diff
-$this->assertEquals($filteredValue, $result['test']);
+$this->assertSame($filteredValue, $result['test']);
```

File: tests/FilterRule/SlugTest.php [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    public function testSlugFilterEmptyIfNotAvailable()
+    {
+        $this->filter->value('slug')->slug('title');
+
+        $result = $this->filter->filter([
+            'test' => 'test',
+        ]);
+
+        $this->assertEquals(['test' => 'test'], $result);
+    }
```
Comment (localheinz) [inline tests/FilterRule/SlugTest.php:68]: @jhuet 

How about introducing an explaining variable here to improve clarity?

``` php
$data = [            
    'test' => 'test',
];

$this->assertSame($data, $this->filter->filter($data));
```

File: tests/FilterRule/SlugTest.php (hunk 3 above)
Comment (jhuet) [inline tests/FilterRule/SlugTest.php]: Oh, ok @localheinz ! Well i guess we can, yes. I honestly copied and passed another test case and reused `assertEquals`. The slug filter should always return a `string` so it'd work too.

File: tests/FilterRule/SlugTest.php (hunk 4 above)
Comment (jhuet) [inline tests/FilterRule/SlugTest.php:68]: Alright, you got it, i'm in a good mood today :)

**Comment by rick-nu** [comment]
Very nice @jhuet, thanks a lot! I'll merge and release a new version :) :+1:

**Comment by rick-nu** [comment]
https://github.com/particle-php/Filter/releases/tag/v1.4.0

**Pull request merged**